matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/OlmMachine.kt

```diff
@@ -0,0 +1,536 @@
+        val adapter = MoshiProvider.providesMoshi().adapter(Event::class.java)
+        val serializedEvent = adapter.toJson(event)
+
+        inner.requestRoomKey(serializedEvent, event.roomId!!)
+    }
+
+    /**
+     * Export all of our room keys.
+     *
+     * @param passphrase The passphrase that should be used to encrypt the key
+     * export.
+     *
+     * @param rounds The number of rounds that should be used when expanding the
+     * passphrase into an key.
+     *
+     * @return the encrypted key export as a bytearray.
+     */
+    @Throws(CryptoStoreErrorException::class)
+    suspend fun exportKeys(passphrase: String, rounds: Int): ByteArray = withContext(Dispatchers.IO) {
+        inner.exportKeys(passphrase, rounds).toByteArray()
+    }
+
+    /**
+     * Import room keys from the given serialized key export.
+     *
+     * @param keys The serialized version of the key export.
+     *
+     * @param passphrase The passphrase that was used to encrypt the key export.
+     *
+     * @param listener A callback that can be used to introspect the
+     * progress of the key import.
+     */
+    @Throws(CryptoStoreErrorException::class)
+    suspend fun importKeys(keys: ByteArray, passphrase: String, listener: ProgressListener?): ImportRoomKeysResult = withContext(Dispatchers.IO) {
+        val decodedKeys = String(keys, Charset.defaultCharset())
+
+        val rustListener = CryptoProgressListener(listener)
+
+        val result = inner.importKeys(decodedKeys, passphrase, rustListener)
+
+        ImportRoomKeysResult(result.total, result.imported)
+    }
+
+    /**
+     * Get a `Device` from the store.
+     *
+     * @param userId The id of the device owner.
+     *
+     * @param deviceId The id of the device itself.
+     *
+     * @return The Device if it found one.
+     */
+    @Throws(CryptoStoreErrorException::class)
+    suspend fun getDevice(userId: String, deviceId: String): CryptoDeviceInfo? = withContext(Dispatchers.IO) {
+        // Our own device isn't part of our store on the rust side, return it
+        // using our ownDevice method
+        if (userId == userId() && deviceId == deviceId()) {
+            ownDevice()
+        } else {
+            val device = inner.getDevice(userId, deviceId)
+            if (device != null) toCryptoDeviceInfo(device) else null
+        }
+    }
+
+    /**
+     * Get all devices of an user.
+     *
+     * @param userId The id of the device owner.
+     *
+     * @return The list of Devices or an empty list if there aren't any.
+     */
+    @Throws(CryptoStoreErrorException::class)
+    suspend fun getUserDevices(userId: String): List<CryptoDeviceInfo> {
+        val ownDevice = ownDevice()
+        val devices = inner.getUserDevices(userId).map { toCryptoDeviceInfo(it) }.toMutableList()
+
+        // EA doesn't differentiate much between our own and other devices of
+        // while the rust-sdk does, append our own device here.
+        devices.add(ownDevice)
```

Comment:
This is strange, you add the current device, event if the `userId` in parameter is not the current user?

Reply:
Right, that's wrong. Nice catch.